### PR TITLE
Add new defender forward charge

### PR DESCRIPTION
**Pull request by psykzz**
## About The Pull Request

Adds a new xeno ability, "forward charge", adds that ability to the defender replacing the headbutt.
Defender now waits 0.5 sec before charging 4 tiles, knocking down and doing 50% dmg to those in the path.

## Why It's Good For The Game

Balacing too many one click stuns on defender

## Changelog
:cl:
add: Forward Charge to defenders, Defender waits 0.5 sec before charging 4 tiles, knocking down and doing 50% dmg to those in the path.
/:cl:

### CODE REVIEW

File: code/modules/mob/living/carbon/xenomorph/castes/defender/abilities_defender.dm [hunk 1]
```diff
@@ -147,6 +147,44 @@
 		X.use_plasma(plasma_cost)
 	add_cooldown()
 
+// ***************************************
+// *********** Forward Charge
+// ***************************************
+/datum/action/xeno_action/activable/forward_charge
+	name = "Forward Charge (80)"
```
Comment (spookydonut) [inline code/modules/mob/living/carbon/xenomorph/castes/defender/abilities_defender.dm]:
```suggestion
	name = "Forward Charge"
```

File: code/modules/mob/living/carbon/xenomorph/castes/defender/abilities_defender.dm [hunk 2]
```diff
@@ -147,6 +147,44 @@
...
+/datum/action/xeno_action/activable/forward_charge/on_cooldown_finish()
+	to_chat(owner, "<span class='xenodanger'>Your exoskeleton quivers as you get ready to use Forward Charge again.</span>")
+	playsound(owner, "sound/effects/xeno_newlarva.ogg", 50, 0, 1)
+	return ..()
+
+/datum/action/xeno_action/activable/forward_charge/use_ability(atom/A)
+	var/mob/living/carbon/Xenomorph/X = owner
+
+	if(!do_after(X, 0.5 SECONDS, FALSE, X, BUSY_ICON_GENERIC))
+		return
```
Comment (spookydonut) [inline code/modules/mob/living/carbon/xenomorph/castes/defender/abilities_defender.dm]:
```suggestion
		return fail_activate()
```

File: code/modules/mob/living/carbon/xenomorph/castes/defender/defender.dm [hunk 3]
```diff
@@ -48,3 +48,21 @@
...
 		set_fortify(FALSE) //Fortify prevents dragging due to the anchor component.
+
+
+// ***************************************
+// *********** Mob overrides
+// ***************************************
+/mob/living/carbon/Xenomorph/Defender/Bump(atom/A)
+	if(!throwing || !throw_source || !thrower) //Must currently be charging to knock aside and hit marines in it's path
+		return ..() //It's not charging; do regular Bump() IE body block but not throw_impact() because defender isn't being thrown
+	if(!ishuman(A)) //Must also be a human; regular Bump() will default to throw_impact() which means ravager will plow through tables but get stopped by cades and walls
```
Comment (spookydonut) [inline code/modules/mob/living/carbon/xenomorph/castes/defender/defender.dm]: dont copy paste comments

File: code/modules/mob/living/carbon/xenomorph/castes/defender/defender.dm [hunk 4]
```diff
@@ -48,3 +48,21 @@
...
+		return ..()
+	var/mob/living/carbon/human/H = A
+	var/extra_dam = rand(xeno_caste.melee_damage_lower, xeno_caste.melee_damage_upper) * 0.5 // 50% dmg reduction
+	H.attack_alien(src, extra_dam, FALSE, TRUE, FALSE, TRUE, INTENT_HARM) //Location is always random, cannot crit, harm only
```
Comment (spookydonut) [inline code/modules/mob/living/carbon/xenomorph/castes/defender/defender.dm]: this is kinda bad.

Comment (psykzz) [reply code/modules/mob/living/carbon/xenomorph/castes/defender/defender.dm]: How come?

Comment (spookydonut) [reply code/modules/mob/living/carbon/xenomorph/castes/defender/defender.dm]: just in general, not a lot that you can do about it.

File: code/modules/mob/living/carbon/xenomorph/castes/defender/abilities_defender.dm [hunk 5]
```diff
@@ -147,6 +147,44 @@
...
 
+// ***************************************
+// *********** Forward Charge
+// ***************************************
+/datum/action/xeno_action/activable/forward_charge
+	name = "Forward Charge"
+	action_icon_state = "charge"
+	mechanics_text = "Charge up to 4 tiles and knockdown any targets in your way."
+	ability_name = "charge"
+	cooldown_timer = DEFENDER_CHARGECOOLDOWN
```
Comment (spookydonut) [inline code/modules/mob/living/carbon/xenomorph/castes/defender/abilities_defender.dm]:
```suggestion
	cooldown_timer = 15 SECONDS
```

File: code/__DEFINES/mobs.dm [hunk 6]
```diff
@@ -545,6 +545,11 @@ GLOBAL_LIST_INIT(xenoupgradetiers, list(XENO_UPGRADE_BASETYPE, XENO_UPGRADE_INVA
 #define DEFENDER_TAILSWIPE_COOLDOWN		12 SECONDS
 #define DEFENDER_FORTIFY_COOLDOWN		1 SECONDS
 #define DEFENDER_CREST_DEFENSE_COOLDOWN	1 SECONDS
+#define DEFENDER_CHARGESPEED			70
+#define DEFENDER_CHARGESTRENGTH			3
+#define DEFENDER_CHARGEDISTANCE			4
+#define DEFENDER_CHARGECOOLDOWN			15 SECONDS
```
Comment (spookydonut) [inline code/__DEFINES/mobs.dm]:
```suggestion
```

File: code/modules/mob/living/carbon/xenomorph/castes/defender/abilities_defender.dm (hunk 5 above)
Comment (psykzz) [reply code/modules/mob/living/carbon/xenomorph/castes/defender/abilities_defender.dm]: Why when everything is done via a define?

Comment (spookydonut) [reply code/modules/mob/living/carbon/xenomorph/castes/defender/abilities_defender.dm]: this is needless obfuscation

File: code/__DEFINES/mobs.dm (hunk 6 above)
Comment (spookydonut) [reply code/__DEFINES/mobs.dm]: this is needless obfuscation

File: code/modules/mob/living/carbon/xenomorph/castes/defender/abilities_defender.dm (hunk 5 above)
Comment (spookydonut) [reply code/modules/mob/living/carbon/xenomorph/castes/defender/abilities_defender.dm]: nothing else is referencing the define.

Comment (Rohesie) [reply code/modules/mob/living/carbon/xenomorph/castes/defender/abilities_defender.dm]: It's the standard we've been following, but we can change it. Spooky's suggestion is better.

File: code/modules/mob/living/carbon/xenomorph/castes/defender/abilities_defender.dm [hunk 7]
```diff
@@ -147,6 +147,44 @@
...
+	plasma_cost = 80
+
+/datum/action/xeno_action/activable/forward_charge/can_use_ability(atom/A, silent = FALSE, override_flags)
+	. = ..()
+	if(!.)
+		return FALSE
+	if(!A)
+		return FALSE
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/xenomorph/castes/defender/abilities_defender.dm]:
```suggestion
		CRASH("forward_charge/can_use_ability invoked without an atom/A")
```
We should be using `atom/movable/AM` here, though. I know other things don't, but, you know.

File: code/modules/mob/living/carbon/xenomorph/castes/defender/abilities_defender.dm [hunk 8]
```diff
@@ -147,6 +147,44 @@
...
+	to_chat(owner, "<span class='xenodanger'>Your exoskeleton quivers as you get ready to use Forward Charge again.</span>")
+	playsound(owner, "sound/effects/xeno_newlarva.ogg", 50, 0, 1)
+	return ..()
+
+/datum/action/xeno_action/activable/forward_charge/use_ability(atom/A)
+	var/mob/living/carbon/Xenomorph/X = owner
+
+	if(!do_after(X, 0.5 SECONDS, FALSE, X, BUSY_ICON_GENERIC))
+		return fail_activate()
+
```
Comment (spookydonut) [inline code/modules/mob/living/carbon/xenomorph/castes/defender/abilities_defender.dm]: Needs a can use ability check here to make sure you aren't stunned etc

**Comment by Rohesie** [comment]
Travis check:
```
code/modules/mob/living/carbon/xenomorph/castes/defender/defender.dm:62:error: xeno_caste.melee_damage_lower: undefined var
```

**Comment by Rohesie** [comment]
Same travis check issue.

File: code/modules/mob/living/carbon/xenomorph/castes/defender/abilities_defender.dm [hunk 9]
```diff
@@ -147,6 +147,44 @@
...
+	if(!A)
+		CRASH("forward_charge/can_use_ability invoked without an atom/A")
+
+/datum/action/xeno_action/activable/forward_charge/on_cooldown_finish()
+	to_chat(owner, "<span class='xenodanger'>Your exoskeleton quivers as you get ready to use Forward Charge again.</span>")
+	playsound(owner, "sound/effects/xeno_newlarva.ogg", 50, 0, 1)
+	return ..()
+
+/datum/action/xeno_action/activable/forward_charge/use_ability(atom/A)
+	var/mob/living/carbon/Xenomorph/X = owner
```
Comment (spookydonut) [inline code/modules/mob/living/carbon/xenomorph/castes/defender/abilities_defender.dm]:
```suggestion
	var/mob/living/carbon/xenomorph/X = owner
```

File: code/modules/mob/living/carbon/xenomorph/castes/defender/abilities_defender.dm [hunk 10]
```diff
@@ -147,6 +147,44 @@
...
+
+	if(!do_after(X, 0.5 SECONDS, FALSE, X, BUSY_ICON_GENERIC, extra_checks = CALLBACK(src, .proc/can_use_ability, A)))
```
Comment (spookydonut) [inline code/modules/mob/living/carbon/xenomorph/castes/defender/abilities_defender.dm]:
```suggestion
	if(!do_after(X, 0.5 SECONDS, FALSE, X, BUSY_ICON_GENERIC, extra_checks = CALLBACK(src, .proc/can_use_ability, A, FALSE, XACT_USE_BUSY)))
```

File: code/modules/mob/living/carbon/xenomorph/castes/defender/defender.dm [hunk 11]
```diff
@@ -48,3 +48,21 @@
 	. = ..()
 	if(stat && fortify)
 		set_fortify(FALSE) //Fortify prevents dragging due to the anchor component.
+
+
+// ***************************************
+// *********** Mob overrides
+// ***************************************
+/mob/living/carbon/Xenomorph/Defender/Bump(atom/A)
```
Comment (spookydonut) [inline code/modules/mob/living/carbon/xenomorph/castes/defender/defender.dm]:
```suggestion
/mob/living/carbon/xenomorph/defender/Bump(atom/A)
```

File: code/modules/mob/living/carbon/xenomorph/castes/defender/abilities_defender.dm [hunk 12]
```diff
@@ -147,6 +147,44 @@
...
+	ability_name = "charge"
+	cooldown_timer = 15 SECONDS
+	plasma_cost = 80
+
+/datum/action/xeno_action/activable/forward_charge/can_use_ability(atom/A, silent = FALSE, override_flags)
+	. = ..()
+	if(!.)
+		return FALSE
+	if(!A)
+		CRASH("forward_charge/can_use_ability invoked without an atom/A")
```
Comment (spookydonut) [inline code/modules/mob/living/carbon/xenomorph/castes/defender/abilities_defender.dm]: shouldn't crash here just return false.

Comment (spookydonut) [reply code/modules/mob/living/carbon/xenomorph/castes/defender/abilities_defender.dm]: also crashing here without at least a . = FALSE will maybe result in can_use_ability returning TRUE

File: code/modules/mob/living/carbon/xenomorph/castes/defender/abilities_defender.dm [hunk 13]
```diff
@@ -147,6 +147,44 @@
...
+// *********** Forward Charge
+// ***************************************
+/datum/action/xeno_action/activable/forward_charge
+	name = "Forward Charge"
+	action_icon_state = "charge"
+	mechanics_text = "Charge up to 4 tiles and knockdown any targets in your way."
+	ability_name = "charge"
+	cooldown_timer = 15 SECONDS
+	plasma_cost = 80
+
```
Comment (spookydonut) [inline code/modules/mob/living/carbon/xenomorph/castes/defender/abilities_defender.dm]:
```suggestion
	use_state_flags = XACT_USE_CRESTED
```
if you want them able to use it while in crest defense mode

**Comment by Rohesie** [comment]
Works good.
It was suggested that non-harm intent should do a tackle attack instead.
Thoughts?

**Comment by Rohesie** [comment]
Let us know when this is ready for another test.

**Comment by psykzz** [comment]
Ready for another test @Rohesie

**Pull request merged**